enterprise/server/util/distributed_client/distributed_client.go

```
@@ -509,64 +509,93 @@ func (c *Proxy) RemoteGetMulti(ctx context.Context, peer string, isolation *dcpb
 }
 
 func (c *Proxy) RemoteReader(ctx context.Context, peer string, r *rspb.ResourceName, offset, limit int64) (io.ReadCloser, error) {
-	isolation := &dcpb.Isolation{
-		CacheType:          r.GetCacheType(),
-		RemoteInstanceName: r.GetInstanceName(),
-	}
-	req := &dcpb.ReadRequest{
-		Isolation: isolation,
-		Key:       digestToKey(r.GetDigest()),
-		Offset:    offset,
-		Limit:     limit,
-		Resource:  r,
-	}
-
 	client, err := c.getClient(ctx, peer)
 	if err != nil {
 		return nil, err
 	}
-
+	req := &dcpb.ReadRequest{
+		Isolation: &dcpb.Isolation{
+			CacheType:          r.GetCacheType(),
+			RemoteInstanceName: r.GetInstanceName(),
+		},
+		Key:      digestToKey(r.GetDigest()),
+		Offset:   offset,
+		Limit:    limit,
+		Resource: r,
+	}
 	stream, err := client.Read(ctx, req)
 	if err != nil {
 		return nil, err
 	}
-	reader, writer := io.Pipe()
+	return newDistributedCacheReader(stream, r.GetDigest().GetSizeBytes() == offset)
```

I am worried that this `r.GetDigest().GetSizeBytes() == offset` would do the wrong thing for non CAS items. I don't think it should happen that any one does an offset read on an AC item, but it feels very sketchy to me to assume that. Have you tested this?

Backing up a bit -- why do we need this expectEOF flag in the first place?

---

I'm pretty sure this is reproducing the same behavior from line 566 in the old version. I think this is the reasoning:
1. If the blob isn't found, we want RemoteReader to return that error, instead of waiting for the resulting Reader to return that error.
2. If we get an EOF from the first stream message, we generally don't want to return an error here, and instead we want to return a Reader that produces zero bytes and EOF. This handles the case where you're reading an empty blob.
3. If we get an EOF from the first stream message, but we expected to read at least some bytes, then this is unexpected and we should return that error right away. 

I'm happy to remove #3, since it seems like it's handling just one subset of the cases where the number of bytes we read is not what we expected.

---

As far as testing it, I ran an app with this change and things seem to be working, but I'm not sure if you had something specific in mind.